jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ContentSourceTransformer.java

```diff
@@ -63,31 +95,22 @@ public Content.Chunk read()
                     LOG.debug("Raw chunk {} {}", rawChunk, this);
                 needsRawRead = rawChunk == null;
                 if (rawChunk == null)
-                    return null;
+                    return afterRead(State.Type.IDLE, null);
             }
 
             if (Content.Chunk.isFailure(rawChunk))
             {
-                Content.Chunk failure = rawChunk;
+                Content.Chunk failureChunk = rawChunk;
                 rawChunk = Content.Chunk.next(rawChunk);
                 needsRawRead = rawChunk == null;
                 if (rawChunk != null)
-                {
-                    finished = true;
-                    release();
-                }
-                return failure;
+                    return afterRead(State.Type.FAILED, failureChunk);
+                return afterRead(State.Type.IDLE, failureChunk);
```

Comment:
I'm not sure this is correct.   You should return the chunk just read, but the failure chunk set to be returned from future reads should be the result of `Content.Chunk.next(rawChunk)`.    If the next chunk is null, then we should return to idle state.

If you had a transitive error, then you would return it forever

Reply:
What you describe is exactly what the code does. The code always returns the chunk just read (`failureChunk`), calls `Chunk.next()` to update `this.rawChunk`, and the logic that follows operates on the "next" chunk.
For transient errors, the code sets `this.rawChunk` to `null`, moves to `IDLE` and returns the transient `failureChunk`. The next read would see `this.rawChunk=null`, `needsRawRead=true` and `IDLE` state.

Reply:
@sbordet On line 107, it sets the afterRead state with `failureChunk` not with the updated `rawChunk`.   The on line 108, it is going to idle with the `failureChunk` instead of null.

 I think it would be clearer if you introduced a `nextChunk` variable.